File: web/src/lib/mail.ts
import nodemailer from "nodemailer";
import type Transporter from "nodemailer/lib/mailer";

type MailConfig = {
  host: string;
  port: number;
  user: string;
  pass: string;
  secure: boolean;
  from: string;
};

let cachedTransport: Transporter | null = null;
let cachedConfigKey: string | null = null;

export function getMailConfig(): MailConfig | null {
  const host = process.env.SMTP_HOST;
  const user = process.env.SMTP_USER;
  const pass = process.env.SMTP_PASSWORD;

  if (!host || !user || !pass) return null;

  return {
    host,
    port: Number(process.env.SMTP_PORT || "587"),
    user,
    pass,
    secure: process.env.SMTP_SECURE === "true",
    from: process.env.EMAIL_FROM || user,
  };
}

export function isSmtpConfigured(): boolean {
  return getMailConfig() !== null;
}

function getTransportConfigKey(config: MailConfig): string {
  return `${config.host}:${config.port}:${config.user}:${config.secure}`;
}

function getSharedTransport(): { transport: Transporter; config: MailConfig } | null {
  const config = getMailConfig();
  if (!config) return null;

  const key = getTransportConfigKey(config);
  if (!cachedTransport || cachedConfigKey !== key) {
    cachedTransport?.close();
    cachedTransport = nodemailer.createTransport({
      pool: true,
      maxConnections: 1,
      maxMessages: 100,
      host: config.host,
      port: config.port,
      secure: config.secure,
      auth: {
        user: config.user,
        pass: config.pass,
      },
    });
    cachedConfigKey = key;
  }

  return { transport: cachedTransport, config };
}

export async function sendVerificationEmail(
  to: string,
  name: string,
  verifyUrl: string
): Promise<void> {
  const shared = getSharedTransport();

  if (!shared) {
    throw new Error("SMTP 설정이 없습니다.");
  }

  const { transport, config } = shared;

  await transport.sendMail({
    from: `"CompliAI" <${config.from}>`,
    to,
    subject: "[CompliAI] 이메일 인증을 완료해 주세요",
    text: [
      `${name}님, CompliAI 회원가입을 환영합니다.`,
      "",
      "아래 링크를 클릭해 이메일 인증을 완료해 주세요.",
      verifyUrl,
      "",
      "링크는 24시간 동안 유효합니다.",
      "본인이 요청하지 않았다면 이 메일을 무시해 주세요.",
    ].join("\n"),
    html: `
      <div style="font-family: sans-serif; max-width: 520px; margin: 0 auto; padding: 24px;">
        <h2 style="color: #6366f1;">CompliAI 이메일 인증</h2>
        <p>${name}님, 회원가입을 환영합니다.</p>
        <p>아래 버튼을 클릭해 이메일 인증을 완료해 주세요.</p>
        <p style="margin: 32px 0;">
          <a href="${verifyUrl}" style="background:#6366f1;color:#fff;padding:12px 24px;border-radius:8px;text-decoration:none;display:inline-block;">
            이메일 인증하기
          </a>
        </p>
        <p style="color:#666;font-size:14px;">버튼이 동작하지 않으면 아래 링크를 복사해 브라우저에 붙여넣으세요.</p>
        <p style="color:#666;font-size:14px;word-break:break-all;">${verifyUrl}</p>
        <p style="color:#999;font-size:12px;">링크는 24시간 동안 유효합니다.</p>
      </div>
    `,
  });
}

export type MonthlyWinnerMailPayload = {
  to: string;
  name: string;
  periodLabel: string;
  winners: Array<{
    name: string;
    score: number;
    praiseCount: number;
    likeCount: number;
  }>;
  rankingsUrl: string;
  isTest?: boolean;
};

function formatWinnersText(
  winners: MonthlyWinnerMailPayload["winners"],
  isTest?: boolean
): string {
  if (winners.length === 0) {
    return isTest
      ? "해당 월 칭찬왕이 없습니다. (실제 자동/수동 발송 시에는 메일이 발송되지 않습니다.)"
      : "해당 월 칭찬왕이 없습니다.";
  }
  return winners
    .map(
      (w, i) =>
        `${winners.length > 1 ? `${i + 1}. ` : ""}${w.name} (${w.score}점 · 칭찬 ${w.praiseCount}건 · 좋아요 ${w.likeCount})`
    )
    .join("\n");
}

function formatWinnersHtml(
  winners: MonthlyWinnerMailPayload["winners"],
  isTest?: boolean
): string {
  if (winners.length === 0) {
    const msg = isTest
      ? "해당 월 칭찬왕이 없습니다. (실제 자동/수동 발송 시에는 메일이 발송되지 않습니다.)"
      : "해당 월 칭찬왕이 없습니다.";
    return `<p style="color:#666;">${msg}</p>`;
  }
  const items = winners
    .map(
      (w) =>
        `<li><strong>${w.name}</strong> — ${w.score}점 (칭찬 ${w.praiseCount}건, 좋아요 ${w.likeCount})</li>`
    )
    .join("");
  return `<ul style="padding-left: 20px;">${items}</ul>`;
}

export async function sendMonthlyWinnerAnnouncement(
  payload: MonthlyWinnerMailPayload
): Promise<void> {
  const shared = getSharedTransport();

  if (!shared) {
    throw new Error("SMTP 설정이 없습니다.");
  }

  const { transport, config } = shared;
  const { to, name, periodLabel, winners, rankingsUrl, isTest } = payload;
  const winnerText = formatWinnersText(winners, isTest);
  const winnerHtml = formatWinnersHtml(winners, isTest);
  const testPrefix = isTest ? "[테스트] " : "";
  const subject =
    winners.length > 0
      ? `${testPrefix}[CompliAI] ${periodLabel} 칭찬왕 발표`
      : `${testPrefix}[CompliAI] ${periodLabel} 칭찬왕 메일 테스트`;

  const introLine =
    winners.length > 0
      ? `${periodLabel} CompliAI 칭찬왕을 발표합니다!`
      : `${periodLabel} 칭찬왕 메일 발송 테스트입니다.`;

  await transport.sendMail({
    from: `"CompliAI" <${config.from}>`,
    to,
    subject,
    text: [
      `${name}님, 안녕하세요.`,
      "",
      introLine,
      "",
      winnerText,
      "",
      `전체 랭킹 보기: ${rankingsUrl}`,
      "",
      isTest ? "※ 관리자 테스트 발송 메일입니다." : "CompliAI와 함께해 주셔서 감사합니다.",
    ].join("\n"),
    html: `
      <div style="font-family: sans-serif; max-width: 520px; margin: 0 auto; padding: 24px;">
        <h2 style="color: #171717;">${testPrefix}${periodLabel} 칭찬왕 🏆</h2>
        <p>${name}님, 안녕하세요.</p>
        <p>${introLine}</p>
        ${winnerHtml}
        <p style="margin: 32px 0;">
          <a href="${rankingsUrl}" style="background:#171717;color:#fff;padding:12px 24px;border-radius:8px;text-decoration:none;display:inline-block;">
            랭킹 보러 가기
          </a>
        </p>
        ${isTest ? '<p style="color:#999;font-size:12px;">※ 관리자 테스트 발송 메일입니다.</p>' : '<p style="color:#666;font-size:14px;">CompliAI와 함께해 주셔서 감사합니다.</p>'}
      </div>
    `,
  });
}

File: api/app/mail.py
import smtplib
import ssl
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText

from app.config import get_settings


def is_smtp_configured() -> bool:
    settings = get_settings()
    return bool(settings.smtp_host and settings.smtp_user and settings.smtp_password)


def _send(to: str, subject: str, text: str, html: str) -> None:
    settings = get_settings()
    if not is_smtp_configured():
        raise RuntimeError("SMTP 설정이 없습니다.")

    msg = MIMEMultipart("alternative")
    msg["Subject"] = subject
    msg["From"] = f'"CompliAI" <{settings.email_from or settings.smtp_user}>'
    msg["To"] = to
    msg.attach(MIMEText(text, "plain", "utf-8"))
    msg.attach(MIMEText(html, "html", "utf-8"))

    if settings.smtp_secure:
        context = ssl.create_default_context()
        with smtplib.SMTP_SSL(settings.smtp_host, settings.smtp_port, context=context) as server:
            server.login(settings.smtp_user, settings.smtp_password)
            server.sendmail(msg["From"], [to], msg.as_string())
        return

    with smtplib.SMTP(settings.smtp_host, settings.smtp_port) as server:
        server.starttls()
        server.login(settings.smtp_user, settings.smtp_password)
        server.sendmail(msg["From"], [to], msg.as_string())


def send_verification_email(to: str, name: str, verify_url: str) -> None:
    text = (
        f"{name}님, CompliAI 회원가입을 환영합니다.\n\n"
        "아래 링크를 클릭해 이메일 인증을 완료해 주세요.\n"
        f"{verify_url}\n\n"
        "링크는 24시간 동안 유효합니다.\n"
        "본인이 요청하지 않았다면 이 메일을 무시해 주세요."
    )
    html = f"""
      <div style="font-family: sans-serif; max-width: 520px; margin: 0 auto; padding: 24px;">
        <h2 style="color: #6366f1;">CompliAI 이메일 인증</h2>
        <p>{name}님, 회원가입을 환영합니다.</p>
        <p>아래 버튼을 클릭해 이메일 인증을 완료해 주세요.</p>
        <p style="margin: 32px 0;">
          <a href="{verify_url}" style="background:#6366f1;color:#fff;padding:12px 24px;border-radius:8px;text-decoration:none;display:inline-block;">
            이메일 인증하기
          </a>
        </p>
        <p style="color:#666;font-size:14px;word-break:break-all;">{verify_url}</p>
        <p style="color:#999;font-size:12px;">링크는 24시간 동안 유효합니다.</p>
      </div>
    """
    _send(to, "[CompliAI] 이메일 인증을 완료해 주세요", text, html)


def send_password_reset_email(to: str, name: str, reset_url: str) -> None:
    text = (
        f"{name}님, 안녕하세요.\n\n"
        "아래 링크를 클릭해 비밀번호를 재설정해 주세요.\n"
        f"{reset_url}\n\n"
        "링크는 1시간 동안 유효합니다."
    )
    html = f"""
      <div style="font-family: sans-serif; max-width: 520px; margin: 0 auto; padding: 24px;">
        <h2 style="color: #6366f1;">CompliAI 비밀번호 재설정</h2>
        <p>{name}님, 안녕하세요.</p>
        <p style="margin: 32px 0;">
          <a href="{reset_url}" style="background:#6366f1;color:#fff;padding:12px 24px;border-radius:8px;text-decoration:none;display:inline-block;">
            비밀번호 재설정하기
          </a>
        </p>
        <p style="color:#666;font-size:14px;word-break:break-all;">{reset_url}</p>
      </div>
    """
    _send(to, "[CompliAI] 비밀번호 재설정", text, html)


def send_monthly_winner_announcement(
    *,
    to: str,
    name: str,
    period_label: str,
    winners: list[dict],
    rankings_url: str,
    is_test: bool = False,
) -> None:
    test_prefix = "[테스트] " if is_test else ""
    if winners:
        winner_text = "\n".join(
            f"{w['name']} ({w['score']}점 · 칭찬 {w['praiseCount']}건 · 좋아요 {w['likeCount']})"
            for w in winners
        )
        winner_html = "<ul>" + "".join(
            f"<li><strong>{w['name']}</strong> — {w['score']}점 (칭찬 {w['praiseCount']}건, 좋아요 {w['likeCount']})</li>"
            for w in winners
        ) + "</ul>"
        subject = f"{test_prefix}[CompliAI] {period_label} 칭찬왕 발표"
        intro = f"{period_label} CompliAI 칭찬왕을 발표합니다!"
    else:
        winner_text = "해당 월 칭찬왕이 없습니다."
        winner_html = f"<p style=\"color:#666;\">{winner_text}</p>"
        subject = f"{test_prefix}[CompliAI] {period_label} 칭찬왕 메일 테스트"
        intro = f"{period_label} 칭찬왕 메일 발송 테스트입니다."

    text = f"{name}님, 안녕하세요.\n\n{intro}\n\n{winner_text}\n\n전체 랭킹 보기: {rankings_url}"
    html = f"""
      <div style="font-family: sans-serif; max-width: 520px; margin: 0 auto; padding: 24px;">
        <h2>{test_prefix}{period_label} 칭찬왕</h2>
        <p>{name}님, 안녕하세요.</p>
        <p>{intro}</p>
        {winner_html}
        <p style="margin: 32px 0;">
          <a href="{rankings_url}" style="background:#171717;color:#fff;padding:12px 24px;border-radius:8px;text-decoration:none;display:inline-block;">
            랭킹 보러 가기
          </a>
        </p>
      </div>
    """
    _send(to, subject, text, html)


def dispatch_verification_email(email: str, name: str, token: str) -> dict:
    from app.config import get_settings
    import os

    settings = get_settings()
    verify_url = f"{settings.app_url}/verify-email?token={token}"
    if is_smtp_configured():
        try:
            send_verification_email(email, name, verify_url)
        except Exception as exc:
            print("[CompliAI] 인증 메일 발송 실패:", email, exc)
        return {
            "sent": True,
            "message": "입력하신 이메일로 인증 링크를 발송 중입니다. 수신까지 1~2분 걸릴 수 있습니다.",
        }
    if os.getenv("NODE_ENV") == "development" or "localhost" in settings.app_url:
        print("[CompliAI] 이메일 인증 링크:", verify_url)
        return {
            "sent": False,
            "verifyUrl": verify_url,
            "message": "SMTP가 설정되지 않아 개발 모드로 동작합니다. 터미널 또는 아래 링크로 인증해 주세요.",
        }
    raise RuntimeError("이메일 발송 설정이 없습니다. 관리자에게 문의해 주세요.")


def dispatch_password_reset_email(email: str, name: str, token: str) -> dict:
    from app.config import get_settings
    import os

    settings = get_settings()
    reset_url = f"{settings.app_url}/reset-password?token={token}"
    if is_smtp_configured():
        try:
            send_password_reset_email(email, name, reset_url)
        except Exception as exc:
            print("[CompliAI] 비밀번호 재설정 메일 발송 실패:", email, exc)
        return {
            "sent": True,
            "message": "입력하신 이메일로 비밀번호 재설정 링크를 발송 중입니다. 수신까지 1~2분 걸릴 수 있습니다.",
        }
    if os.getenv("NODE_ENV") == "development" or "localhost" in settings.app_url:
        print("[CompliAI] 비밀번호 재설정 링크:", reset_url)
        return {
            "sent": False,
            "resetUrl": reset_url,
            "message": "SMTP가 설정되지 않아 개발 모드로 동작합니다. 터미널 또는 아래 링크로 재설정해 주세요.",
        }
    raise RuntimeError("이메일 발송 설정이 없습니다. 관리자에게 문의해 주세요.")
